### PR TITLE
router(refactor): single navigation stream

### PR DESCRIPTION
**Pull request by jasonaden**
This is a major refactor of how the router previously worked. There are a couple major advantages of this refactor, and future work will be built on top of it.

First, we will no longer have multiple navigations running at the same time. Previously, a new navigation wouldn't cause the old navigation to be cancelled and cleaned up. Instead, multiple navigations could be going at once, and we imperatively checked that we were operating on the most current `router.navigationId` as we progressed through the Observable streams. This had some major faults, the biggest of which was async races where an ongoing async action could result in a redirect once the async action completed, but there was no way to guarantee there weren't also other redirects that would be queued up by other async actions. After this refactor, there's a single Observable stream that will get cleaned up each time a new navigation is requested.

Additionally, the individual pieces of routing have been pulled out into their own operators. While this was needed in order to create one continuous stream, it also will allow future improvements to the testing APIs as things such as Guards or Resolvers should now be able to be tested in much more isolation.

* Add the new `router.transitions` observable of the new `NavigationTransition` type to contain the transition information
* Update `router.navigations` to pipe off of `router.transitions`
* Re-write navigation Observable flow to a single configured stream
* Refactor `switchMap` instead of the previous `mergeMap` to ensure new navigations cause a cancellation and cleanup of already running navigations
* Wire in existing error and cancellation logic so cancellation matches previous behavior

### CODE REVIEW

**Comment by jasonaden** [comment]
[Presubmit](http://test/OCL:212209294:BASE:212209308:1536549443912:f89479b1)

**Comment by jasonaden** [comment]
[Global presubmit](https://test.corp.google.com/OCL:212209294:BASE:212583121:1536730294929:e53217fb)

**Review by benlesh — CHANGES_REQUESTED** [review]

File: packages/router/src/router.ts [hunk 1]
```diff
@@ -326,9 +342,226 @@ export class Router {
...
+          if (this.urlUpdateStrategy === 'deferred' && !t.extras.skipLocationChange) {
+            this.setBrowserUrl(this.rawUrlTree, !!t.extras.replaceUrl, t.id);
+          }
+        }),
+
+        activate(
+            this.rootContexts, this.routeReuseStrategy,
+            (evt: Event) => this.triggerEvent(evt)),
+
+        tap({next: () => completed = true, complete: () => completed = true}),
```
[the remark this answers is not in the document]
Comment (benlesh) [reply packages/router/src/router.ts]: Technically, you could also just define one function above that did it:

```ts
const setCompleted = () => completed = true;

tap({
  next: setCompleted,
  complete: setCompleted,
});
```

It's a minor thing... hence "nit"

File: packages/router/src/router.ts [hunk 2]
```diff
@@ -576,24 +815,22 @@ export class Router {
...
-            this.executeScheduledNavigation(nav);
-            // a failed navigation should not stop the router from processing
-            // further navigations => the catch
-            return nav.promise.catch(() => {});
-          } else {
-            return <any>of (null);
-          }
-        }))
-        .subscribe(() => {});
+    this.navigations.subscribe(
```
[the remark this answers is not in the document]
Comment (alfaproject) [reply packages/router/src/router.ts]: Offtopic: the recommendation is for readability or performance, or?

File: packages/router/src/router.ts [hunk 3]
```diff
@@ -326,9 +342,226 @@ export class Router {
...
+                t.id, this.serializeUrl(t.extractedUrl), e));
+            try {
+              t.resolve(this.errorHandler(e));
+            } catch (ee) {
+              t.reject(ee);
+            }
+          }
+          return EMPTY;
+        }), );
+      })) as any as Observable<NavigationTransition>;
```
[the remark this answers is not in the document]
Comment (jasonaden) [reply packages/router/src/router.ts]: current `g3` requirement that will go away when TS is updated. Added a comment to remove this in the future.

File: packages/router/src/router.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasonaden) [reply packages/router/src/router.ts]: @alfaproject It was for readability only.

**Review by benlesh — CHANGES_REQUESTED** [review]

**Review by benlesh — APPROVED** [review]

**Review by alxhub — APPROVED** [review]

**Pull request closed without merging**